### PR TITLE
[diem-framework] specs for NetworkIdentity

### PR DESCRIPTION
**Pull request by mengxu-fb**
This commit adds the specifications for the `NetworkIdentity` module and
changes a few implementation pieces in order to keep the SMT solver
happy.

The `NetworkIdentity` should be fully specified now. An interesting part
in this code is to simulate two set operations, set union and set
difference, via vector operations. For this, two generic functions are
added, `add_members_internal` and `remove_members_internal` and they
might be useful outside the context of `NetworkIdentity` as well.

The verification takes 40-50 seconds on my MBP, which isn't good.
(and is slightly above the 40 seconds default timeout limit...). But
considering the amount of loop invariants and extensive use of quantifiers,
this is not a totally unreasonable number.

**EDIT**: took 95 seconds in CI time to verify this module. The following
entry can be found in the log of the CI runs --- Yes, this module can still
verify because of the `pragma timeout = 120`.

```
PASS [  95.008s]                                   move-prover::testsuite  prover diem[default]::modules/NetworkIdentity.move
```

Given that the tests in CI is executed in parallel, this is not a very concerning
number and does not add too much to the overall CI time. But please let me
know if I am too optimistic about this.

Only the first commit needs to be reviewed. The second commit updates
the auto-generated doc and `.mv` file.

## Motivation

Add specs

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

CI will cover the verification of this `NetworkIdentity.move`

### CODE REVIEW

File: language/diem-framework/modules/NetworkIdentity.move [hunk 1]
```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
...
-            Vector::swap_remove(identities, i);
+        let i = 0;
+        while ({
+            spec {
+                assert i <= num_to_add;
+                // the set can never reduce in size
+                assert len(members) >= len(old(members));
+                // the current set maintains the uniqueness of the elements
+                assert forall j in 0..len(members), k in 0..len(members): members[j] == members[k] ==> j == k;
+                // the left-split of the current set is exactly the same as the origianl set
```
Comment (gregnazario) [inline language/diem-framework/modules/NetworkIdentity.move:187]: *original

File: language/diem-framework/modules/NetworkIdentity.move [hunk 2]
```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
...
+    // =================================================================
+    // Set operation simulation
+
+    /// Add all elements that appear in `to_add` into `members`.
+    ///
+    /// The `members` argument is essentially a set simulated by a vector, hence
+    /// the uniqueness of its elements are guaranteed, before and after the bulk
+    /// insertion. The `to_add` argument, on the other hand, does not guarantee
+    /// to be a set and hence can have duplicated elements.
+    fun add_members_internal<T: copy>(
```
Comment (gregnazario) [inline language/diem-framework/modules/NetworkIdentity.move:172]: We should probably just make a `Set` type somewhere in the language, since now it's separated out :)

File: language/diem-framework/modules/NetworkIdentity.move [hunk 3]
```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
...
+        to_remove: &vector<T>,
+    ): bool {
+        let num_existing = Vector::length(members);
+        let num_to_remove = Vector::length(to_remove);
 
-    spec remove_identity {
+        let i = 0;
+        while ({
+            spec {
+                assert i <= num_to_remove;
```
Comment (gregnazario) [inline language/diem-framework/modules/NetworkIdentity.move:251]: I'm assuming it's `<=` because it comes before the condition

File: language/diem-framework/modules/NetworkIdentity.move [hunk 4]
```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
...
+            old_members: old(members),
+            new_members: members,
+        };
+        // ensures that the `members` argument is and remains a set
+        invariant forall i in 0..len(members), j in 0..len(members):
+            members[i] == members[j] ==> i == j;
+        // returns whether an element is removed from the set
+        ensures result == (exists i in 0..len(to_remove): contains(old(members), to_remove[i]));
+    }
+    spec schema RemoveMembersInternalEnsures<T: drop> {
```
Comment (gregnazario) [inline language/diem-framework/modules/NetworkIdentity.move:294]: Can you not use these in the loop invariants?

File: language/diem-framework/modules/NetworkIdentity.move [hunk 5]
```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
...
+        // TODO(mengxu): this is to force the prover to honor the "opaque" pragma in the ignore opaque setting
+        aborts_if [abstract] false;
+        aborts_if [concrete] false;
+        include RemoveMembersInternalEnsures<T> {
+            old_members: old(members),
+            new_members: members,
+        };
+        // ensures that the `members` argument is and remains a set
+        invariant forall i in 0..len(members), j in 0..len(members):
+            members[i] == members[j] ==> i == j;
```
Comment (gregnazario) [inline language/diem-framework/modules/NetworkIdentity.move:290]: Just because we use this in like 4 places, maybe we should make this a `spec schema` to have an invariant for set uniqueness?

**Review by gregnazario — COMMENTED** [review]
Looks good, I'll defer to someone who knows the spec language a little better than me, but I think it all makes sense

File: language/diem-framework/modules/NetworkIdentity.move (hunk 2 above)
Comment (mengxu-fb) [reply language/diem-framework/modules/NetworkIdentity.move:172]: Yes, we are actually discussing this (we have a thread on the move-prover channel).

File: language/diem-framework/modules/NetworkIdentity.move (hunk 3 above)
Comment (mengxu-fb) [reply language/diem-framework/modules/NetworkIdentity.move:251]: Yes.

A **loop invariant** is a property of a program loop that is true at every iteration **before** hitting the loop body. And the loop breaking condition (`i < num_to_remove`) is considered as part of the loop body.

File: language/diem-framework/modules/NetworkIdentity.move (hunk 4 above)
Comment (mengxu-fb) [reply language/diem-framework/modules/NetworkIdentity.move:294]: Actually we can, I guess I did not use it just because of convention...

File: language/diem-framework/modules/NetworkIdentity.move (hunk 5 above)
Comment (mengxu-fb) [reply language/diem-framework/modules/NetworkIdentity.move:290]: Good point!

File: language/diem-framework/modules/NetworkIdentity.move (hunk 4 above)
Comment (gregnazario) [reply language/diem-framework/modules/NetworkIdentity.move:294]: Okay interesting, just figure why write it twice

File: language/diem-framework/modules/NetworkIdentity.move [hunk 6]
```diff
@@ -44,93 +46,91 @@ module DiemFramework::NetworkIdentity {
...
-        ensures exists<NetworkIdentity>(account_addr);
         modifies global<NetworkIdentity>(account_addr);
     }
 
     /// Return the underlying `NetworkIdentity` bytes
     public fun get(account_addr: address): vector<vector<u8>> acquires NetworkIdentity {
-        assert(exists<NetworkIdentity>(account_addr), ENETWORK_ID_DOESNT_EXIST);
+        assert(
+            exists<NetworkIdentity>(account_addr),
+            Errors::invalid_state(ENETWORK_ID_DOESNT_EXIST)
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:56]: I think we generally use Errors::not_published or similar, but they can be also use cases where invalid state is more appropriate.

File: language/diem-framework/modules/NetworkIdentity.move [hunk 7]
```diff
@@ -44,93 +46,91 @@ module DiemFramework::NetworkIdentity {
...
-            true
+        let prior_identities = if (exists<NetworkIdentity>(account_addr)) {
+            global<NetworkIdentity>(account_addr).identities
         } else {
-            false
-        }
-    }
+            vec()
+        };
+        let has_change = (exists i in 0..len(to_add): !contains(prior_identities, to_add[i]));
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:96]: Instead of 0..len(vec) you can use range(vec). Or even better quantify over the elements in the vector:

`let has_change = (exists x in to_add: !contains(prior_identities, x)`

File: language/diem-framework/modules/NetworkIdentity.move [hunk 8]
```diff
@@ -44,93 +46,91 @@ module DiemFramework::NetworkIdentity {
...
 
-    spec add_identity {
-        ensures contains<vector<u8>>(identities, to_add);
+        let post handle = global<NetworkIdentity>(account_addr).identity_change_events;
+        let post msg = NetworkIdentityChangeNotification {
+            identities: global<NetworkIdentity>(account_addr).identities,
+            time_rotated_seconds: DiemTimestamp::spec_now_seconds(),
+        };
+
+        aborts_if len(to_add) == 0;
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:104]: May want `aborts_if len(to_add) with Errors::INVALID_ARGUMENT`. We specify the category. Same for other abort specs.

File: language/diem-framework/modules/NetworkIdentity.move [hunk 9]
```diff
@@ -140,25 +140,177 @@ module DiemFramework::NetworkIdentity {
...
+            old_members: prior_identities,
+            new_members: global<NetworkIdentity>(account_addr).identities,
+        };
         modifies global<NetworkIdentity>(account_addr);
-        invariant exists<NetworkIdentity>(account_addr);
-        invariant num_identities <= MAX_ADDR_IDENTITIES;
+        emits msg to handle if has_change;
+    }
+
+    // =================================================================
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:164]: I'm wondering whether we should move this to the Vector module for reuse as long as we do not have sets? Can also be done in a separate PR.

@sblackshear @tzakian

File: language/diem-framework/modules/NetworkIdentity.move [hunk 10]
```diff
@@ -140,25 +140,177 @@ module DiemFramework::NetworkIdentity {
...
+        members: &mut vector<T>,
+        to_add: &vector<T>,
+    ): bool {
+        let num_to_add = Vector::length(to_add);
+        let num_existing = Vector::length(members);
+
+        let i = 0;
+        while ({
+            spec {
+                invariant i <= num_to_add;
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:183]: Quite a bummer that this is so complex. Good to see we can verify it.

File: language/diem-framework/modules/NetworkIdentity.move [hunk 11]
```diff
@@ -140,25 +140,177 @@ module DiemFramework::NetworkIdentity {
...
+            };
+            i = i + 1;
+        };
+
+        Vector::length(members) > num_existing
+    }
+    spec add_members_internal {
+        pragma opaque;
+        // TODO(mengxu): this is to force the prover to honor the "opaque" pragma in the ignore opaque setting
+        ensures [concrete] true;
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:212]: Yeah, functions with loops should probably always be opaque. For now this is a good workaround.

File: language/diem-framework/modules/NetworkIdentity.move [hunk 12]
```diff
@@ -140,25 +140,177 @@ module DiemFramework::NetworkIdentity {
...
+        include UniqueMembers<T>;
+        // returns whether a new element is added to the set
+        ensures result == (exists i in 0..len(to_add): !contains(old(members), to_add[i]));
+    }
+    spec schema AddMembersInternalEnsures<T: copy> {
+        old_members: vector<T>;
+        new_members: vector<T>;
+        to_add: vector<T>;
+        // everything in the `to_add` vector must be in the updated set
+        ensures forall i in 0..len(to_add): contains(new_members, to_add[i]);
```
Comment (wrwg) [inline language/diem-framework/modules/NetworkIdentity.move:229]: Some of those quantifiers might be simplified by directly using `x in vec`.

**Review by wrwg — CHANGES_REQUESTED** [review]

**Comment by mengxu-fb** [comment]
The recent commits simplifies the `forall/exists i in 0..len(vec): P(vec[i])`;
with the notion that iterates over all elements of a vector: `forall/exists e in vec: P(e);`.
This improves the readability of the spec.

But the weird thing is... it takes even more time to verify `remove_members_internal` (originally: 90 seconds, now: 220 seconds)... If we further apply the simplification to the loop invariants, the verification could not finish within 1000 seconds (I lost patience after that...) @wrwg

**Comment by wrwg** [comment]
> The recent commits simplifies the `forall/exists i in 0..len(vec): P(vec[i])`;
> with the notion that iterates over all elements of a vector: `forall/exists e in vec: P(e);`.
> This improves the readability of the spec.
> 
> But the weird thing is... it takes even more time to verify `remove_members_internal` (originally: 90 seconds, now: 220 seconds)... If we further apply the simplification to the loop invariants, the verification could not finish within 1000 seconds (I lost patience after that...) @wrwg

The resulting Boogie encoding should be pretty much the same; if not, it can be fixed. Perhaps submit as is (with verification turned off and a TODO) and I can investigate.

**Review by wrwg — DISMISSED** [review]

**Comment by mengxu-fb** [comment]
Thanks Wolfgang. The verification is turned-off now and a TODO is there to help locate the problem.

/land
